oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,569 @@
+    ///
+    /// # Errors
+    ///
+    /// Returns [`QuantileError::InsufficientSampleSize`] if the sample size
+    /// is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oximeter::Quantile;
+    /// let mut q = Quantile::new(0.5).unwrap();
+    /// for o in 1..=100 {
+    ///    q.append(o).unwrap();
+    /// }
+    /// assert_eq!(q.estimate().unwrap(), 50.0);
+    /// ```
+    pub fn estimate(&self) -> Result<f64, QuantileError> {
+        if self.is_empty() {
+            return Err(QuantileError::InsufficientSampleSize);
+        }
+
+        if self.len() >= FILLED_MARKER_LEN as u64 {
+            return Ok(self.marker_heights[2]);
+        }
+
+        // Try to find an index in heights that is correlated with the p value
+        // when we have less than 5 samples, but more than 0.
+        let mut heights = self.marker_heights;
+        float_ord::sort(&mut heights);
+        let idx = (heights.len() as f64 - 1.) * self.p();
+        return Ok(heights[idx.round() as usize]);
+    }
+
+    /// Append a value/observation to the population and adjust the heights.
+    ///
+    /// This comprises steps B.1, B.2, B.3 (adjust heights) in the P² algorithm,
+    /// including finding the cell k containing the input value and updating the
+    /// current and desired marker positions.
+    ///
+    /// Returns an empty result or an error.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`QuantileError::NonFiniteValue`] if the value is not finite
+    /// when casting to a float.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oximeter::Quantile;
+    /// let mut q = Quantile::new(0.9).unwrap();
+    /// q.append(10).unwrap();
+    /// assert_eq!(q.len(), 1);
+    /// ```
+    pub fn append<T>(&mut self, value: T) -> Result<(), QuantileError>
+    where
+        T: HistogramSupport,
+    {
+        if !value.is_finite() {
+            return Err(QuantileError::NonFiniteValue);
+        }
+        // We've already checked that the value is finite.
+        let value_f = value.to_f64().unwrap();
+
+        // if !self.is_filled {
+        if self.len() < FILLED_MARKER_LEN as u64 {
+            self.marker_heights[self.len() as usize] = value_f;
+            self.marker_positions[4] += 1;
+            if self.len() == FILLED_MARKER_LEN as u64 {
+                float_ord::sort(&mut self.marker_heights);
+                self.adaptive_init();
+            }
+            return Ok(());
+        }
+
+        // Find the cell k containing the new value.
+        let k = match self.find_cell(value_f) {
+            Some(4) => {
+                self.marker_heights[4] = value_f;
+                3
+            }
+            Some(i) => i,
+            None => {
+                self.marker_heights[0] = value_f;
+                0
+            }
+        };
+
+        // Handle rounding issues as described in
+        // <https://aakinshin.net/posts/p2-quantile-estimator-rounding-issue>.
```

Comment:
Nice catch by you and this person! The continual summing will add more error as the sums get much larger than the desired increments. Nice.